[WebDriver][BiDi] Support event subscription id's

lauromoura opened the pull request:
#### 4bc81ef1ff541444927e336548af2ef1cc5808c5
<pre>
[WebDriver][BiDi] Support event subscription id&apos;s
<a href="https://bugs.webkit.org/show_bug.cgi?id=287929">https://bugs.webkit.org/show_bug.cgi?id=287929</a>

Reviewed by BJ Burg.

Each call to BiDi session.subscribe should return an unique UUID to
represent that specific subscription. And in turn, `session.unsubscribe`
should allow unsubscribing either by ID or by the event name.

This commit changes the current event system to better match the spec
and keep a list of subscriptions, instead of the current model of
flagging individual events as enabled or not.

Note: As the &quot;params&quot; context to `session.unsubscribe` is deprecated,
we&apos;ll ignore it for now as we support only global events. But
subscribing to specific contexts is still defined in the spec and we&apos;ll
add in the future.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::eventIsEnabled): Process the list of subscriptions.
(WebDriver::Session::subscribeForEvents): Create and store subscriptions,
returning the new UUID for each call.
(WebDriver::Session::unsubscribeByIDs): Added.
(WebDriver::Session::unsubscribeByEventName): Added.
(WebDriver::Session::enableGlobalEvent): Deleted.
(WebDriver::Session::disableGlobalEvent): Deleted.
* Source/WebDriver/Session.h: Added Subscription struct.
(WebDriver::Subscription::isGlobal const): Added.
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::bidiSessionSubscribe): Refactor, calling
the new methods in Session.
(WebDriver::WebDriverService::bidiSessionUnsubscribe): Refactor, calling
the new methods in Session.
* WebDriverTests/TestExpectations.json: Updated test results after this
  change.

Canonical link: <a href="https://commits.webkit.org/295276@main">https://commits.webkit.org/295276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/383958155594d11314bca31a777615475f2a32a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104635 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14763 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109847 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55306 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24734 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32890 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79458 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19243 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94433 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59770 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12507 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54679 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112238 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31796 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88542 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88161 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22453 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33063 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10830 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31723 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37065 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31516 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33075 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+    for (const auto& subscription : m_subscriptions) {
 
-    return m_globalEventSet.contains(eventName);
+        // FIXME Add support to subscribe to specific browsing contexts
```
Style nit: `FIXME:`

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
-    HashSet<String> topLevelBrowsingContexts;
-    // FIXME Add support to subscribe to specific browsing contexts
-    // https://bugs.webkit.org/show_bug.cgi?id=282981
+    for (const auto& subscription : m_subscriptions) {
```
Prefer to call this `m_eventSubscriptions`

burg replied:
This seems super inefficient in the case where we have a lot of "cache misses". Can we keep a HashSet of String/AtomicString of event names for which there are *any* subscriptions and use that as an initial filter?

lauromoura replied:
I went with a `HashMap<eventname, count>`, as the same event name might be subscribed to multiple times (e.g. global + different contexts)

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+
+void Session::subscribe(Vector<String>&& events, Vector<String>&& browsingContextIDs, Vector<String>&& userContextIDs, Function<void(CommandResult&&)>&& completionHandler)
+{
+    // TODO Process/validate list of event names (e.g. expanding if given only the module name)
```
I'd change this to a FIXME, typically we leave FIXME's in but not TODO's.

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+    // https://bugs.webkit.org/show_bug.cgi?id=291371
+    auto subscriptionID = WTF::createVersion4UUIDString();
+    Subscription subscription = { subscriptionID, WTFMove(events), WTFMove(browsingContextIDs), WTFMove(userContextIDs) };
+    m_subscriptions.add(subscriptionID, WTFMove(subscription));
```
You might be able to inline the initializer for `subscription` since we just used the ID after this point.

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+    return false;
+}
+
+void Session::subscribe(Vector<String>&& events, Vector<String>&& browsingContextIDs, Vector<String>&& userContextIDs, Function<void(CommandResult&&)>&& completionHandler)
```
Nit: `subscribeForEvents`

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
 }
 
-void Session::enableGlobalEvent(const String& eventName)
+void Session::unsubscribeByIDs(Vector<String>&& subscriptionIDs, Function<void(CommandResult&&)>&& completionHandler)
```
Naming is a little awkward since it is not explicit about what type of ID it is. What about `unsubscribeForEvents` to match `subscribeForEvents`?

lauromoura replied:
Instead of renaming the method, I used `using SubscriptionID = String`. Or would it be better to actually rename it?

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,89 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+    // Only modify the actual subscriptions after we validated all requested events.
+    m_subscriptions = WTFMove(subscriptionsToKeep);
+
+    // TODO Send the unsubscription request to the browser?
```
Not sure if you meant to leave in this comment.

lauromoura replied:
Oops. Removed

burg commented on Source/WebDriver/Session.h:
```diff
@@ -47,6 +47,18 @@ namespace WebDriver {
 class CommandResult;
 class SessionHost;
 
+struct Subscription {
```
Naming suggestion: `EventSubscription`

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2732,32 +2733,79 @@ void WebDriverService::bidiSessionSubscribe(unsigned id, RefPtr<JSON::Object>&&p
+            completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidArgument, std::nullopt, errorMessage, id));
+            return;
+        }
+        auto subscriptionObj = JSON::Object::create();
```
Naming: call this `result`

lauromoura replied:
Renamed, and renamed the `completionHandler` param from `result` to `subscriptionResult`.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2732,32 +2733,79 @@ void WebDriverService::bidiSessionSubscribe(unsigned id, RefPtr<JSON::Object>&&p
+        subscriptionObj->setString("subscription"_s, result.result()->asString());
 
-    completionHandler(WebSocketMessageHandler::Message::reply("success"_s, id, JSON::Value::null()));
+        completionHandler(WebSocketMessageHandler::Message::reply("success"_s, id, WTFMove(subscriptionObj)));
```
Drive-by feedback: it's a little weird that the first argument to `WebSocketMessageHandler::Message::reply` is a string rather than an enum. Is there a reason to do it this way?

I think it's implied that it's a success, since it's calling `reply()` and not `fail()` right?

lauromoura replied:
The whole `Message` thing will be refactored soon (e.g. storing the json payload instead of the serialized one, to make the relay cleaner, as commented in https://github.com/WebKit/WebKit/pull/42968#discussion_r2057344462), so I'll keep it in mind.

burg commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2721,9 +2722,9 @@ void WebDriverService::bidiSessionStatus(unsigned id, RefPtr<JSON::Object>&&, Fu
 {
     // https://w3c.github.io/webdriver-bidi/#command-session-subscribe
-    auto eventNames = parameters->getArray("events"_s);
+    auto eventNamesJS = parameters->getArray("events"_s);
```
Nit: technically t's not JS, it's an untyped JSON array (`JSON::Array<JSON::Value>`)

burg reviewed (approved):
This is looking good. I'd like to see a slightly more efficient `Session::eventIsEnabled` using a hash set of all subscribed event names. This seems like the most common way to subscribe to events- by name and optionally by browsing context or user context.

burg commented on Source/WebDriver/Session.cpp:
```diff
@@ -3282,21 +3283,119 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+                continue;
+            }
+
+            // FIXME Add support to subscribe to specific browsing contexts
```
Style nit: I usually write FIXME with : (FIXME:). And prefer TODOs removed or converted to FIXMEs.

lauromoura replied:
Oops, I fixed the other one and missed this.

burg commented on Source/WebDriver/Session.h:
```diff
@@ -273,7 +292,8 @@ public RefCounted<Session>
 
     // https://w3c.github.io/webdriver-bidi/#events
-    HashSet<String> m_globalEventSet;
+    HashMap<AtomString, unsigned> m_eventSubscriptionCount;
```
Naming: m_eventSubscriptionCount*s* ?

burg reviewed (approved):
This looks great.

Please update the commit message with the new test expectations. Some of them were obviously related to this change (session.subscribe) while others seem to be gardening. I'd prefer to only check in relevant expectations changes with the code changes. Unreviewed test rebaselines can happen at any time on trunk/main.

burg commented on WebDriverTests/TestExpectations.json:
```diff
@@ -2031,7 +2039,7 @@
             },
             "test_create_remove_contexts": {
-                "expected": { "all": { "status": ["FAIL"], "bug": "webkit.org/b/288107"}}
+                "expected": { "all": { "status": ["FAIL"], "bug": "webkit.org/b/288107", "note": "False PASS, Error on the teardown fixture, given we don't support user contexts."}}
```
Seems like we need a separate gardening commit just to update user context related results?

lauromoura replied:
All these gardening entries are related to this subscription ID patch, even if not under `subscribe/` or `unsubscribe`. Usually they were tests that either:

- Check if a given event was not emitted: Most of there are now passing, as `session.subscribe` return the subscription ID.
- Check if a given event was emitted: These now are timeout instead of fail, as we likely don't implement the event yet.

Regarding this specific entry, the test body itself is passing, thus the "unexpected pass" result, but if I mark as  `["PASS"]` or `["FAIL", "PASS"]`, the teardown step fails and reports an error.  I'll try to run the WPT tests directly (outside webkit) to see if it's something on the fixtures or on our tooling wrapping the expectations.

lauromoura committed:
[WebDriver][BiDi] Support event subscription id's
https://bugs.webkit.org/show_bug.cgi?id=287929

Reviewed by BJ Burg.

Each call to BiDi session.subscribe should return an unique UUID to
represent that specific subscription. And in turn, `session.unsubscribe`
should allow unsubscribing either by ID or by the event name.

This commit changes the current event system to better match the spec
and keep a list of subscriptions, instead of the current model of
flagging individual events as enabled or not.

Note: As the "params" context to `session.unsubscribe` is deprecated,
we'll ignore it for now as we support only global events. But
subscribing to specific contexts is still defined in the spec and we'll
add in the future.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::eventIsEnabled): Process the list of subscriptions.
(WebDriver::Session::subscribeForEvents): Create and store subscriptions,
returning the new UUID for each call.
(WebDriver::Session::unsubscribeByIDs): Added.
(WebDriver::Session::unsubscribeByEventName): Added.
(WebDriver::Session::enableGlobalEvent): Deleted.
(WebDriver::Session::disableGlobalEvent): Deleted.
* Source/WebDriver/Session.h: Added Subscription struct.
(WebDriver::Subscription::isGlobal const): Added.
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::bidiSessionSubscribe): Refactor, calling
the new methods in Session.
(WebDriver::WebDriverService::bidiSessionUnsubscribe): Refactor, calling
the new methods in Session.
* WebDriverTests/TestExpectations.json: Updated test results after this
  change.

Canonical link: https://commits.webkit.org/295276@main
